python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,445 @@
+"""Definitions of types related to coordinate systems."""
+
+import abc
+import collections.abc
+from typing import Optional, Sequence, Tuple, Union
+
+import attrs
+import numpy as np
+import numpy.typing as npt
+
+from .types import str_or_seq_length
+from .types import to_string_tuple
+
+
+@attrs.define(frozen=True)
+class Axis:
+    """A description of an axis of a coordinate system
+
+    Lifecycle: experimental
+    """
+
+    name: str
+    """Name of the axis."""
+    unit: Optional[str] = None
+    """Optional string name for the units of the axis."""
+
+
+@attrs.define(frozen=True)
+class CoordinateSpace(collections.abc.Sequence):
+    """A coordinate space for spatial data.
+
+    Args:
+        axes: The axes of the coordinate system in order.
+
+    Lifecycle: experimental
+    """
+
+    axes: Tuple[Axis, ...] = attrs.field(converter=tuple)
+
+    @axes.validator
+    def _validate(self, _, axes: Tuple[Axis, ...]) -> None:
+        if not axes:
+            raise ValueError("at least one")
+        if len(set(axis.name for axis in self.axes)) != len(axes):
+            raise ValueError("unique")
+
+    def __len__(self) -> int:
+        return len(self.axes)
+
+    def __getitem__(self, index: int) -> Axis:  # type: ignore[override]
```

Comment:
doing the `Sequence[Axis]` change above means this `type: ignore` should be able to go away.